### PR TITLE
Add option to run browser in background

### PR DESCRIPTION
**Pull request by dennisschagt**
Implements idea from https://github.com/newsboat/newsboat/pull/1259#issuecomment-739348017.

We normally ignore queued input when the browser process finishes to avoid https://github.com/newsboat/newsboat/issues/63.
This is necessary because we change between terminal modes (using `Stfl::reset()`).

This PR makes it possible to run the browser in the background without changing terminal modes.
In that way, we don't need to ignore queued input.

Marking as draft as there are a few things to be done:
- [x] Decide if this (boolean config option) is an acceptable way to implement this -> Minoru prefers a separate operation
- [x] Document new setting -> Documented the new operation
- [x] ~~Get exit code from `run_program()`?~~ -> Using different function which does return the program's exit code
- [x] ~~Use both values of `"browser-run-in-background"` option in `OPEN_IN_BROWSER` tests~~ -> Duplicated test

### CODE REVIEW

**Comment by dennisschagt** [comment]
> Decide if this is an acceptable way to implement this

A small issue I see with this solution is that the new option is not synchronized with the `browser` configuration.
Users might want to run an interactive browser for some items while running other programs in the background (like the `download` program in https://github.com/newsboat/newsboat/pull/1259#issuecomment-739348017).

We might be able to do some processing on the `browser` configuration variable (to detect a string at the start like `--run-in-background`).
However, I rather keep the current implementation of this PR (with the `browser-run-in-background` as a separate option) and think it is easier to understand.

Somewhat related, if we implement an open-with command (https://github.com/newsboat/newsboat/issues/1171), it might be useful to add an `in-background` argument.

**Comment by Minoru** [comment]
I was just writing a comment pointing out the same issue :) I think a better way to resolve it is to add `open-in-background-browser` *operation*, so users can choose the behaviour on a case-by-case basis. This would also provide an easier-to-understand replacement for `browser "firefox &"` [that our FAQ advertises](https://newsboat.org/releases/2.21/docs/faq.html#_after_i_open_a_link_in_the_browser_newsboat_stays_unresponsive_until_the_browser_is_closed).

But I'd really like to hold off any implementations until we're sure we understood tonycpus's problem. If it's indeed just slowness of `open-in-browser`, then we'll discuss this PR further. But it might be something else entirely, and we won't know until tonycpus describe their workflow and provide a bit more detail.

**Comment by tonycpsu** [comment]
@Minoru : My use case is invoking external programs like `youtube-dl` on a number of articles in a feed.  For years, I've used the trick outlined [here](https://github.com/newsboat/newsboat/issues/599#issuecomment-518317592) to be able to do this, with my normal browser set to an interactive one.  My current macro looks like this:

    macro . set browser "~/bin/download %u" ; open-in-browser ; set browser "elinks"; toggle-article-read "read";

I apologize for neglecting to include the `toggle-article-read` bit in my previous comment in the other issue.

Executing this sequence would open the "browser", with a brief clearing of the screen switching from the newsboat UI to the command-line, then the "browser" would do its thing while newsboat's UI continued to run.  (NB: The browser was technically not in the background in this case, as I would see error messages from it occasionally, which is kind of a bug since it messed up the newsboat UI, but also kind of a feature since it gave me a visual indication that something went wrong with the download.). The downloaded article was marked read, the input focus moved to the next item, and I could quickly repeat the sequence to queue another.  

Re: the solution in this PR, it's not working as well for me.  It does queue the keypresses, but seems to wait for the "browser" to finish its job on the first one before advancing to and downloading the next article.  (I tried adding `&` to the command to no avail.)

Seconding the suggestion of a separate command, as I sometimes do want to run an interactive browser.

**Comment by Minoru** [comment]
Thanks for the answers!

> toggle-article-read "read"

Wow, I didn't even know one can do that ._. I though `toogle-…` only *toggles*. TIL, thanks :) I'll file an issue to update the docs.

> NB: The "browser" was technically not in the background in this case, as I would see error messages from it occasionally

Okay, so ~/bin/download is a script that does something like `youtube-dl --quiet $1 &` — it spawns `youtube-dl` in the background and suppresses almost all output, but the remaining output (like errors) can still obtrude Newsboat's screen. I think if we implement an option or operation for this, we'd want to always redirect all output to /dev/null, because it makes no sense to let a background process (that we no longer control) to write onto our screen. If the user wants to see that ouput, they would have to redirect it themselves by adding `>` to their `browser` setting.

> seems to wait for the "browser" to finish its job on the first one (with the newsboat UI frozen) before advancing to and downloading the next article. (I tried adding & to the command to no avail.)

I couldn't reproduce this, but I also don't think it's important at the moment — if @dennisschagt agrees to turn this setting into an operation, the code will change and will need to be tested anew.

**Comment by dennisschagt** [comment]
> tonycpsu: the solution in this PR, it's not working as well for me. It does queue the keypresses, but seems to wait for the "browser" to finish its job on the first one (with the newsboat UI frozen) before advancing to and downloading the next article. (I tried adding & to the command to no avail.)

This happened because Newsboat was waiting for the `stdout` pipe to be closed.
As the program (`youtube-dl`?) inherited that handle, it kept Newsboat waiting.
The current implementation does not wait for any of the `stdout`/`stderr` handles so adding `&` lets Newsboat continue directly.

> Minoru: I think if we implement an option or operation for this, we'd want to always redirect all output to /dev/null, because it makes no sense to let a background process (that we no longer control) to write onto our screen. If the user wants to see that ouput, they would have to redirect it themselves by adding > to their browser setting.

Sounds good, the current implementation makes sure the output of running programs is not displayed at all.

> Minoru: I think a better way to resolve it is to add open-in-background-browser operation, so users can choose the behaviour on a case-by-case basis.

Sounds good. I now use the name `open-in-noninteractive-browser` but, if you prefer `open-in-background-browser`, I would be fine with that.

> Minoru: This would also provide an easier-to-understand replacement for browser "firefox &" that our FAQ advertises.

The current implementation still requires the ampersand if the user wants to continue directly (with the program continuing in the background).

I have 3 reasons to keep this operation blocking:
- It is easier to test the operation (testing if a program has run has inherent racing issues if we don't wait for it to finish)
- Running a program in a non-blocking way with a blocking implementation is easy (by running in background using &).
    Going the other way is hard (or even impossible). If a user would want to wait for the program to finish, they would have no way if we implement the operation in a non-blocking way.
- This operation now works like the regular `open-in-browser` operation. This makes it easier for users to debug issues with a "browser" configuration (they can temporarily change to the interactive version of the operation to view the output while everything else still works the same).

File: src/feedlistformaction.cpp [hunk 1]
```diff
@@ -215,7 +215,20 @@ bool FeedListFormAction::process_operation(Operation op,
...
 			if (feed) {
-				return open_feed_in_browser(feed);
+				const bool interactive = true;
+				return open_feed_in_browser(feed, interactive);
+			}
+		} else {
+			v->show_error(_("No feed selected!"));
+		}
+		break;
+	case OP_OPENINBROWSER_NONINTERACTIVE:
```
Comment (Minoru) [inline src/feedlistformaction.cpp:225]: Handlers for `OP_OPENINBROWSER` and `OP_OPENINBROWSER_NONINTERACTIVE` only differ by a single line (`const bool interactive`). Can they be extracted into a separate method that just takes `interactive` as an argument?

Same applies to all other dialogs that handle those operations.

If this requires too many changes, or intrusive changes that risk introducing bugs, I'd prefer to extract this work into a separate PR, to be merged after the 2.22 release.

File: src/itemlistformaction.cpp [hunk 2]
```diff
@@ -156,7 +157,22 @@ bool ItemListFormAction::process_operation(Operation op,
...
+				return false;
+			}
+			invalidate(itempos);
+		} else {
+			v->show_error(_("No item selected!"));
+		}
+	}
+	break;
+	case OP_OPENINBROWSER_NONINTERACTIVE: {
+		LOG(Level::INFO, "ItemListFormAction: opening item at pos `%u'", itempos);
```
Comment (Minoru) [inline src/itemlistformaction.cpp:171]: I think it'd be nice if the logging statements mentioned if they refer to interactive or non-interactive operation. It's just a log, so I think it'd be fine to do something like `"interactive: %u", static_cast<int>(interactive)`.

**Review by Minoru — CHANGES_REQUESTED** [review]
This looks nice! Thanks for the work.

> I now use the name open-in-noninteractive-browser but, if you prefer open-in-background-browser, I would be fine with that. 

No, I think it's really better to highlight non-interactivity — and I agree with your rationale for making this operation blocking, so calling it "background" wouldn't even be true anymore.

One bikeshed comment I *do* have is that the new name kind of implies the existence of something called "noninteractive browser", and the users might go looking for `noninteractive-browser` setting and so on. I think naming the operation `open-in-browser-noninteractively` would alleviate this. What do you think?

**Comment by dennisschagt** [comment]
> One bikeshed comment I do have is that the new name kind of implies the existence of something called "noninteractive browser", and the users might go looking for noninteractive-browser setting and so on. I think naming the operation open-in-browser-noninteractively would alleviate this. What do you think?

Sounds good, applied in https://github.com/newsboat/newsboat/pull/1335/commits/56a42eeac3ed45ec06df8bcd3d4889c898ed947d

File: doc/keycmds.dsv [hunk 3]
```diff
@@ -14,6 +14,7 @@ next||J||Jump to next list entry.
 prev||K||Jump to previous list entry.
 random-unread||^K||Jump to a random unread article.
 open-in-browser||o||Use browser to open the URL associated with the current article, feed, or entry in the URL view.
+open-in-noninteractive-browser||o||Use browser to open the URL associated with the current article, feed, or entry in the URL view. This operation works similar to `open-in-browser`, but the output of the browser (stdout and stderr) are not shown.
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply doc/keycmds.dsv]: Nice suggestion, applied it and included some example usage of the new operation in https://github.com/newsboat/newsboat/pull/1335/commits/2a6fe14615198b3ae8a39b6c78440c8926b92b8f

File: src/feedlistformaction.cpp (hunk 1 above)
Comment (dennisschagt) [reply src/feedlistformaction.cpp]: > Can they be extracted into a separate method that just takes interactive as an argument?

Yes, done.

> If this requires too many changes, or intrusive changes that risk introducing bugs, I'd prefer to extract this work into a separate PR, to be merged after the 2.22 release.

I think the changes are relatively save.
However, I'm fine if you want to drop them from this PR (and merge them after the 2.22 release).

File: src/itemlistformaction.cpp (hunk 2 above)
Comment (dennisschagt) [reply src/itemlistformaction.cpp]: Done:
- https://github.com/newsboat/newsboat/pull/1335/commits/ed6991651d579bd8d5fd9fcd397e82f3ec3f292f#diff-d5a1b4371967aac8adde72ce6f97d087e4f8d305f7176749ecdaf22d2ad62844R761
- https://github.com/newsboat/newsboat/pull/1335/commits/80f1766b060b0b2d6d7cb64737cd4ca01f4d2270#diff-1551378fbd86d2c7065b34517b3d6746dc86e657464942650be7078863458be3R547

File: src/keymap.cpp [hunk 4]
```diff
@@ -148,6 +148,13 @@ static const std::vector<OpDesc> opdescs = {
 		_("Open URL of article, feed, or entry in URL view"),
 		KM_FEEDLIST | KM_ARTICLELIST | KM_ARTICLE | KM_URLVIEW
 	},
+	{
+		OP_OPENINBROWSER_NONINTERACTIVE,
+		"open-in-browser-noninteractively",
+		"",
+		_("Open URL of article, feed, or entry in non-interactive browser"),
```
Comment (Minoru) [inline src/keymap.cpp:155]: I think this should've reflected the change too:

```suggestion
		_("Open URL of article, feed, or entry in a browser, non-interactively"),
```

**Review by Minoru — APPROVED** [review]
Awesome, thanks for the work! This introduces a new string though ("Running browser"), plus two fuzzies, so I'll hold off the merge until after the 2.22 release.

(When I decided that I'll do a "call for translations" two weeks prior to releases, it wasn't my intent to create a "freeze period" of any sort—but looks like that's what ended up happening. At this point, I'm just rolling with it :) Let's see if it helps us avoid hotfix releases.)

**Comment by tonycpsu** [comment]
This works perfectly!  Thanks to both of you for the work on this!

File: src/keymap.cpp (hunk 4 above)
Comment (dennisschagt) [reply src/keymap.cpp]: Done: https://github.com/newsboat/newsboat/compare/80f1766b060b0b2d6d7cb64737cd4ca01f4d2270..7f6ce535a715779e5ff937e6ff56bda4c7831c18

**Comment by dennisschagt** [comment]
Sounds good, thanks for the reviews!

**Pull request merged**